docs/config.bouncer.md

# Prepare config for the app

Copy `env.dist` as `.env` and edit its content environmental variables.

## Register bouncer in Crowdsec LAPI

Generate a bouncer API key following [CrowdSec documentation](https://doc.crowdsec.net/docs/cscli/cscli_bouncers_add),
get a bouncer API key from your CrowdSec with a command

```shell
cscli bouncers add mikrotik-bouncer
```

Copy the API key printed. You **WILL NOT** be able the get it again.
Paste this API key as the value for bouncer environment variable `CROWDSEC_BOUNCER_API_KEY`

## Set firewall rules to control

In previous steps you can get the required firewall rules which will be used to
be updated.

Adjust other variables in .env file as needed, especially host to MikroTik
device and CrowdSec endpoint. See section below.

## Configuration options

The bouncer configuration is made via environment variables.

TODO: use golang docstring generator to list env vars and settings

### CROWDSEC_BOUNCER_API_KEY

`CROWDSEC_BOUNCER_API_KEY` - default value: unset, required,
CrowdSec bouncer API key required to be authorized to request local API.

### CROWDSEC_URL

`CROWDSEC_URL` - default value: `http://crowdsec:8080/`, required,
Host and port of CrowdSec LAPI agent.

### CROWDSEC_ORIGINS

`CROWDSEC_ORIGINS` - default value: unset, optional,
Space separated list of CrowdSec origins to filter from LAPI,
in example `crowdsec cscli`.

### DEBUG_DECISIONS_MAX

`DEBUG_DECISIONS_MAX` - default value: `-1`, optional,
Set number of decisions to process at max, useful for debugging.
Set to 3 to make things less spammy.

### LOG_LEVEL

`LOG_LEVEL` - default value: `1`, optional,
Minimum log level for bouncer in [zerolog levels](https://pkg.go.dev/github.com/rs/zerolog#readme-leveled-logging)

### LOG_FORMAT_JSON

`LOG_FORMAT_JSON` - default value: `true`, optional,
Use logs in JSON format, set to `false` for plain text zerolog format
with key=value, useful only in certain debug sessions

### MIKROTIK_HOST

`MIKROTIK_HOST` - default value: unset, required,
MikroTik device address to access RouterOS API ( `ip:port`)

### MIKROTIK_USER

`MIKROTIK_USER` - default value: unset, required,
Mikrotik device username to access RouterOS API

### MIKROTIK_PASS

`MIKROTIK_PASS` - default value: unset, required,
Mikrotik device password to access RouterOS API

### MIKROTIK_TLS

`MIKROTIK_TLS` -  default value: `true`, optional,
User TLS to connect to MikroTik API,

### MIKROTIK_IPV4

`MIKROTIK_IPV4` - default value: `true`, optional,
IPv4 support, set to `true` to enable processing IPv4 blocklists

### MIKROTIK_IPV6

`MIKROTIK_IPV6` - default value: `true`,  optional,
IPv6 support, set to `true` to enable processing IPv6 blocklists

### MIKROTIK_FIREWALL_FILTER_ENABLE

`MIKROTIK_FIREWALL_FILTER_ENABLE` - default value: `true`, optional,
enable updating firewall filter rules (filter input, forward, output)
See [Firewall Filter or Raw](config.mikrotik.md#firewall---filter-or-raw) for more details.

### IP_FIREWALL_FILTER_RULES_SRC

`IP_FIREWALL_FILTER_RULES_SRC` - default value: unset, required if `MIKROTIK_IPV4` is set to true,
comma separated numbers of IPv4 firewall filter rules to update on access-list change,
and to set src-address-list in it,
those are created during configuration, for example `1,2` (filter input, forward, output)

### IP_FIREWALL_FILTER_RULES_DST

`IP_FIREWALL_FILTER_RULES_DST` - default value: unset, required if `MIKROTIK_IPV4` is set to true,
comma separated numbers of IPv4 firewall filter rules to update on access-list change,
and to set dst-address-list in it,
those are created during configuration, for example `3,4` (filter input, forward, output)

### IPV6_FIREWALL_FILTER_RULES_SRC

`IPV6_FIREWALL_FILTER_RULES_SRC` - default value: unset, required if `MIKROTIK_IPV6` is set to true,
comma separated numbers of IPv6 firewall filter rules to update on access-list change,
and to set src-address-list in it,
those are created during configuration , for example `0,1` (filter input, forward, output)

### IPV6_FIREWALL_FILTER_RULES_DST

`IPV6_FIREWALL_FILTER_RULES_DST` - default value: unset, required if `MIKROTIK_IPV6` is set to true,
comma separated numbers of IPv6 firewall filter rules to update on access-list change,
and to set dst-address-list in it,
those are created during configuration , for example `2,3` (filter input, forward, output)

### MIKROTIK_FIREWALL_RAW_ENABLE

`MIKROTIK_FIREWALL_RAW_ENABLE` - default value: `true`, optional,
enable updating firewall raw rules (raw prerouting, output).
See [Firewall Filter or Raw](config.mikrotik.md#firewall---filter-or-raw) for more details.

### IP_FIREWALL_RAW_RULES_SRC

`IP_FIREWALL_RAW_RULES_SRC` - default value: unset, required if `MIKROTIK_IPV4` is set to true,
comma separated numbers of IPv4 firewall raw rules to update on access-list change,
and to set src-address-list in it,
those are created during configuration, for example `1` (raw prerouting, output)

### IP_FIREWALL_RAW_RULES_DST

`IP_FIREWALL_RAW_RULES_DST` - default value: unset, required if `MIKROTIK_IPV4` is set to true,
comma separated numbers of IPv4 firewall raw rules to update on access-list change,
and to set dst-address-list in it,
those are created during configuration, for example `2` (raw prerouting, output)

### IPV6_FIREWALL_RAW_RULES_SRC

`IPV6_FIREWALL_RAW_RULES_SRC` - default value: unset, required if `MIKROTIK_IPV6` is set to true,
comma separated numbers of IPv6 firewall raw rules to update on access-list change,
and to set src-address-list in it,
those are created during configuration , for example `0` (raw prerouting, output)

### IPV6_FIREWALL_RAW_RULES_DST

`IPV6_FIREWALL_RAW_RULES_DST` - default value: unset, required if `MIKROTIK_IPV6` is set to true,
comma separated numbers of IPv6 firewall raw rules to update on access-list change,
and to set dst-address-list in it,
those are created during configuration , for example `1` (raw prerouting, output)

### MIKROTIK_ADDRESS_LIST

`MIKROTIK_ADDRESS_LIST` - default value: `crowdsec`, optional,
prefix for target address-list in MikroTik device, no special chars,
no spaces etc, generated name will be with a timestamp suffix,
if you set it to `crowdsec` then access-list will be named as
`crowdsec_2025-05-19_15-01-09` or something like it (local time),

### MIKROTIK_TIMEOUT

`MIKROTIK_TIMEOUT` - default value: `10s`, optional,
set timeout when trying to connect to the MikroTik,
recommended to keep it under `60s`

### MIKROTIK_UPDATE_FREQUENCY

`MIKROTIK_UPDATE_FREQUENCY` - default value: `1h`, optional,
Set default frequency to update MikroTik address-lists and firewall rules.
This is useful if you have very low [DEFAULT_TTL_MAX](#default_ttl_max) value.

This is useful if you have disabled [TRIGGER_ON_UPDATE](#trigger_on_update),
or enabled `USE_MAX_TTL=true` and set [DEFAULT_TTL_MAX](#default_ttl_max).

### USE_MAX_TTL

`USE_MAX_TTL` - default value: `false`, optional,
Set to `true` if you want to truncate timeout for the address in address-list
so that your address lists expire faster

### DEFAULT_TTL_MAX

`DEFAULT_TTL_MAX` - default value: `4h`, optional,
If [USE_MAX_TTL](#use_max_ttl) is `true`  and new address timeout is above
[DEFAULT_TTL_MAX](#default_ttl_max) then that address will have timeout
set to [USE_MAX_TTL](#use_max_ttl) value.

For example new decision comes in, and address should be banned for 4 days,
but `DEFAULT_TTL_MAX=4h` will make it to be added with `timeout=4h`.
Notice that the original 4 day ban will be respected in the application cache
or from incoming CrowdSec decisions, but on MikroTik it will have 4h.

Yet it is good to quickly expire old address-lists automatically, because
new ones will come in with refreshed entries for the same address IPs to block.

Because CrowdSec publishes new lists at least once an hour then that address
will be readded to the new list every hour until expires.

This helps to avoid having thousands addresses in hundreds address-lists in
the MikroTik.

Recommended value is at least 2x longer than the frequency you get updates from
the CrowdSec, so on basic setup 4h should be sufficient. For locations with
possible network disruptions 8h or 16 would be recommended
(but then why ban if there is no internet? :) ).

For weaker/older devices it may be better to keep it really low like 2h.

If you get frequent updates from your CrowdSec LAPI (say every 5 minutes),
and you have [TRIGGER_ON_UPDATE=true](#trigger_on_update), then
you could even set it to as low as 15min.

Must be longer than [MIKROTIK_UPDATE_FREQUENCY](#mikrotik_update_frequency).

### TRIGGER_ON_UPDATE

`TRIGGER_ON_UPDATE` - default value: `true`, optional,
if you set it to true, then trigger MikroTik address-list and firewall update
immediately (look at [TICKER_INTERVAL](#ticker_interval)).

This makes ban added from other tools being applied faster, but for the
price of creating new address-list and firewall update.
Effectively you really want this enabled, and also have [USE_MAX_TTL](#use_max_ttl)
set to say 4h for quicker old address expiration.

If set to `false` then the address will not be banned until the next loop
of the [MIKROTIK_UPDATE_FREQUENCY](#mikrotik_update_frequency) is executed,
so on default settings it may take up to 1h before address is banned.

### TICKER_INTERVAL

`TICKER_INTERVAL` - default value: `10s`, optional
how frequently process streamed decisions from CrowdSec LAPI,

Notice this is a golang [time.Duration](https://pkg.go.dev/time#Duration)
format, but the value cannot be equal or less than `0s`.

This will vary depending on the current length of the IP addresses
to be blocked - so for example if you test with 4k addresses inserted and it
takes 10s then adding 20k addresses may take more (let say 25s).

The best if this is the total time used to update MikroTik
address lists and firewall plus about 30% just to prevent bouncer stuck waiting
to acquire lock for the update.

If you get frequent delays in acquiring lock then try to increase this value
higher, certain devices are quite slow and they need at least `30s` or `60s`
for processing.

Sometimes it is just better to buy better faster hardware.

### GOMAXPROCS

`GOMAXPROCS` - default value: unset (automatic number of processors), optional,
Set default processes to use by golang app, especially useful to prevent it
from getting excessively throttled in the containers,
Recommended value `1`.

### METRICS_ADDRESS

`METRICS_ADDRESS` - default value: `:2112`, optional,
Address to use to start metrics server in Prometheus format, metrics are
exposed under `/metrics` path, without authorization (not implemented).

### TZ

`TZ` - default value: unset, optional,
set desired timezone, otherwise if empty then it will take local time from
the machine it runs on. It affects logging and name of the address-list
suffix created on the MikroTik device. Example `UTC` or `Europe/Warsaw`.


docs/config.mikrotik.md
# MikroTik config

## Enable MikroTik RouterOS API

See [MikroTik RouterOS API](https://help.mikrotik.com/docs/spaces/ROS/pages/47579160/API)

## MikroTik user

Add user to MikroTik to allow access via RouterOS API.

```shell
/user group add name=crowdsec policy=api,read,write
/user add name=crowdsec-bouncer-user password=hunter2 group=crowdsec disabled=no
```

Remember to filter out access for the created user for given address only etc.

## Optional - Firewall rules with Log or Passthrough

To avoid unexpected issues it is advised to change the new rules to `passthrough`
so that you can see the number of packets that got through it.

Additional logging should help in checking if the setup works as expected.

If it works then change rules to `drop` one by one to make it actually workign
as expected :)

Below commands assume no logging and drop as default action.

## Firewall - IPv6 or IPv4

If you have dual-stacik (IPv6 and IPV4) available publicly, then enable both.
If you dont have one then it's better to use the only one you have, because
it speeds up device configuration and lowers consumed resources such as memory.

## Firewall - filter or raw

There is a difference between `firewall filter` and `firewall raw`, they serve
different purposes and require different computing powers - in short `filter`
is computationally heavier than `raw`, but allows doing more advanced traffic
filtering such as connection tracking and layer 7 protocol detection.

- raw drops packets as early as possible, and they are stateless
- raw is processed before filter rules, so it can save resources such as cpu
  and memory, especially if you are hitting device limits
- filter works after some other rules, like connection tracking
  and thus they are stateful and allow better manipulation of certain use cases
- using raw rules only should be done only when dealing with specific use cases
  such as DDOS (crowdsec is not really good at it, though)

Which one to choose?

- by default both are enabled, and first `raw` will drop most of the traffic,
  thus `filter` shoudl have much less to process and thus not consume that much
  resources if `raw` is not enabled.

- if the MikroTik device still struggles such as hitting limits of the cpu/memory
  of the device then try to use `raw` mode only.

## Creating Firewall filter rules

### Creating IPv6 firewall filter rules

For IPv6 - create IPv6 'drop' filter rules in `input` and `forward`
chain with the source address list set to `crowdsec` at the top.

Below are snippets to use, make sure to replace `ether1` with your desired interface.
Notice that if you use `place-before=0` then the order below is important,
and for `dst-address-list` we do not define interface.

```shell
/ipv6 firewall filter \
add action=drop dst-address-list=crowdsec chain=input \
place-before=0 comment="crowdsec input drop rules - dst"

/ipv6 firewall filter \
add action=drop dst-address-list=crowdsec chain=forward \
place-before=0 comment="crowdsec forward drop rules - dst"

/ipv6 firewall filter \
add action=drop src-address-list=crowdsec chain=input \
in-interface=ether1 \
place-before=0 comment="crowdsec input drop rules - src"

/ipv6 firewall filter \
add action=drop src-address-list=crowdsec chain=forward \
in-interface=ether1 \
place-before=0 comment="crowdsec forward drop rules - src"

```

The best would be to add them just after default `bad_ipv6` rules.

### Creating IPv4 firewall rules

For IPv4 - create IP `drop` filter rules in `input` and `forward` chain with the
source address list set to `crowdsec` at the top or just before
generic packet counter rule.

Below are snippets to use, make sure to replace `ether1` with your desired interface,
assuming that rule 0 is a dummy passthrough for packet counting added by default
to MikroTik, and rule 1 is whatever but we want to insert CrowdSec before it.
Notice that if you use `place-before=1` then the order below is important,
and for `dst-address-list` we do not define interface.

```shell
/ip firewall filter \
add action=drop dst-address-list=crowdsec chain=input \
place-before=1 comment="crowdsec input drop rules - dst"

/ip firewall filter \
add action=drop dst-address-list=crowdsec chain=forward \
place-before=1 comment="crowdsec forward drop rules - dst"

/ip firewall filter \
add action=drop src-address-list=crowdsec chain=input \
in-interface=ether1 \
place-before=1 comment="crowdsec input drop rules - src"

/ip firewall filter \
add action=drop src-address-list=crowdsec chain=forward \
in-interface=ether1 \
place-before=1 comment="crowdsec forward drop rules - src"

```

### List firewall filter rules

Get the list of firewall rules which were added, this will be needed later.

```shell
/ip firewall filter print without-paging

/ipv6 firewall filter print without-paging
```

Write down numbers of the rules on the most left column.

For example for IPv4:

```shell
> /ip firewall filter print without-paging

Flags: X - disabled, I - invalid; D - dynamic
 0  D ;;; special dummy rule to show fasttrack counters
      chain=forward action=passthrough

 1    ;;; crowdsec forward drop rules - src
      chain=forward action=drop src-address-list=crowdsec in-interface=ether1 log=no log-prefix=""

 2    ;;; crowdsec input drop rules - src
      chain=input action=drop src-address-list=crowdsec in-interface=ether1 log=no log-prefix=""

 3    ;;; crowdsec forward drop rules - dst
      chain=forward action=drop dst-address-list=crowdsec log=no log-prefix=""

 4    ;;; crowdsec input drop rules - dst
      chain=input action=drop dst-address-list=crowdsec log=no log-prefix=""

 5    ;;; defconf: accept established,related,untracked
      chain=input action=accept connection-state=established,related,untracked

 6    ;;; defconf: drop invalid
      chain=input action=drop connection-state=invalid

```

then:

- `IP_FIREWALL_FILTER_RULES_SRC` would be `1,2`
- `IP_FIREWALL_FILTER_RULES_DST` would be `3,4`

Similar, for IPv6:

```shell
> /ipv6 firewall filter print without-paging
Flags: X - disabled, I - invalid; D - dynamic
 0    ;;; crowdsec input drop rules - src
      chain=input action=drop src-address-list=crowdsec in-interface=ether1 log=no log-prefix=""

 1    ;;; crowdsec forward drop rules - src
      chain=forward action=drop src-address-list=crowdsec in-interface=ether1 log=no log-prefix=""

 2    ;;; crowdsec input drop rules - dst
      chain=input action=drop log=no log-prefix=""

 3    ;;; crowdsec forward drop rules - dst
      chain=forward action=drop log=no log-prefix=""

 4    ;;; defconf: drop invalid
      chain=input action=drop connection-state=invalid

 5    ;;; defconf: accept established,related,untracked
      chain=input action=accept connection-state=established,related,untracked
```

then:

- `IPV6_FIREWALL_FILTER_RULES_SRC` would be `0,1`
- `IPV6_FIREWALL_FILTER_RULES_DST` would be `2,3`

## Creating firewall raw rules

### Creating IPv6 firewall raw rules

For IPv6 - create IPv6 'drop' filter rules in `prerouting` and `output`
chain with the source/destination address list set to `crowdsec` at the top.

Below are snippets to use, make sure to replace `ether1` with your desired interface.
Notice that if you use `place-before=0` then the order below is important.

```shell
/ipv6 firewall raw \
add action=drop src-address-list=crowdsec chain=prerouting \
in-interface=ether1 \
comment="crowdsec prerouting drop rules - src"

/ipv6 firewall raw \
add action=drop dst-address-list=crowdsec chain=prerouting \
comment="crowdsec prerouting drop rules - dst"

/ipv6 firewall raw \
add action=drop dst-address-list=crowdsec chain=output \
comment="crowdsec output drop rules - dst"

```

The best would be to add them just after default `bad_ipv6` rules.

### Creating IPv4 firewall raw rules

For IPv4 - create IP `drop` filter rules in `prerouting` and `output` chain with the
source/destination address list set to `crowdsec` at the top or just before
generic packet counter rule.

Below are snippets to use, make sure to replace `ether1` with your desired interface,
assuming that rule 0 is a dummy passthrough for packet counting added by default
to MikroTik, and rule 1 is whatever but we want to insert CrowdSec before it.
Notice that if you use `place-before=1` then the order below is important.

```shell
/ip firewall raw \
add action=drop src-address-list=crowdsec chain=prerouting \
in-interface=ether1 \
comment="crowdsec prerouting drop rules - src"

/ip firewall raw \
add action=drop dst-address-list=crowdsec chain=prerouting \
comment="crowdsec prerouting drop rules - dst"

/ip firewall raw \
add action=drop dst-address-list=crowdsec chain=output \
comment="crowdsec output drop rules - dst"

```

### List firewall raw rules

Get the list of firewall rules which were added, this will be needed later.

```shell
/ip firewall raw print without-paging

/ipv6 firewall raw print without-paging
```

Write down numbers of the rules on the most left column.

For example for IPv4:

```shell
> /ip firewall raw print without-paging
Flags: X - disabled, I - invalid; D - dynamic
 0  D ;;; special dummy rule to show fasttrack counters
      chain=prerouting action=passthrough

 1    ;;; crowdsec prerouting drop rules - src
      chain=prerouting action=drop log=no log-prefix="" src-address-list=crowdsec_2025-07-06_12-50-38

 2    ;;; crowdsec output drop rules - dst
      chain=output action=drop log=no log-prefix="" dst-address-list=crowdsec_2025-07-06_12-50-38

 3    ;;; crowdsec prerouting drop rules - dst
      chain=prerouting action=drop log=no log-prefix="" dst-address-list=crowdsec

```

then:

- `IP_FIREWALL_RAW_RULES_DST` would be `2,3` (output,prerouting)
- `IP_FIREWALL_RAW_RULES_SRC` would be `1` (prerouting)

Similar, for IPv6:

```shell
> /ipv6 firewall raw print without-paging
Flags: X - disabled, I - invalid; D - dynamic
 0    ;;; crowdsec prerouting drop rules - src
      chain=prerouting action=drop in-interface=ether1 src-address-list=crowdsec

 1    ;;; crowdsec prerouting drop rules - dst
      chain=prerouting action=drop dst-address-list=crowdsec

 2    ;;; crowdsec output drop rules - dst
      chain=output action=drop dst-address-list=crowdsec


```

then:

- `IPV6_FIREWALL_RAW_RULES_DST` would be `1,2` (output,prerouting)
- `IPV6_FIREWALL_RAW_RULES_SRC` would be `0` (prerouting)


config.go
package main

import (
	"fmt"
	"maps"
	"os"
	"regexp"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"

	"github.com/spf13/viper"
)

var (
	addressList           string   // mikrotik filter address-list prefix
	crowdsecBouncerAPIKey string   // crowdsec bouncer API key
	crowdsecBouncerURL    string   // url to crowdsec lapi
	crowdsecOrigins       []string // CORS

	// use this for coding debug sessions only
	// max decisions to use when processing,
	// set to low as 3 to enable, thus limiting number of items processed
	// default -1, means process everything
	debugDecisionsMax int

	// default time for items without TTL,
	// this is required to allow automatic list expiry, if unsure set it to 1d
	defaultTTL time.Duration

	// default time for items with TTL, this is used to truncate incoming TTL
	// to specific value to prevent of address-list with addresses
	// which would expire after few days
	// default 24h
	// we assume that we get updates from crowdsec at least once per hour
	// so 1 day seems reasonable for now
	maxTTL time.Duration

	// set to true if you want to use maxTTL
	useMaxTTL bool

	enableFirewallFilter bool   // enable updating firewall filter rules
	srcFilterRuleIdsIPv4 string // comma separated firewall filter rule ids for IPv4 for source rules
	srcFilterRuleIdsIPv6 string // comma separated firewall filter rule ids for IPv6 for source rules
	dstFilterRuleIdsIPv4 string // comma separated firewall filter rule ids for IPv4 for destination rules
	dstFilterRuleIdsIPv6 string // comma separated firewall filter rule ids for IPv6 for destination rules

	enableFirewallRaw bool   // enable updating firewall raw rules
	srcRawRuleIdsIPv4 string // comma separated firewall raw rule ids for IPv4 for source rules
	srcRawRuleIdsIPv6 string // comma separated firewall raw rule ids for IPv6 for source rules
	dstRawRuleIdsIPv4 string // comma separated firewall raw rule ids for IPv4 for destination rules
	dstRawRuleIdsIPv6 string // comma separated firewall raw rule ids for IPv6 for destination rules

	logLevel    string // 0=debug, 1=info
	metricsAddr string // prometheus listen address

	mikrotikHost string        // address of the mikrotik device
	password     string        // mikrotik api password
	timeout      time.Duration //mikrotik command timeout duration
	useIPV4      bool          // set to true to process IPv4 addresses
	useIPV6      bool          // set to true to process IPv6 addresses
	username     string        // mikrotik api username
	useTLS       bool          // use TLS in communication with mikrotik

	// run mikrotik address-list+fw update on received decision event
	// defaults to true if you want faster blocking/unblocking
	triggerOnUpdate bool
	//mikrotik update frequency to process create new address list and update firewall
	updateFreq time.Duration

	// how frequently process streamed decisions from CrowdSec LAPI
	// the best if this is as close to the total time used to update MikroTik
	// address lists and firewall as possible
	// if you get frequent delays in acquiring lock then try to increase this value
	tickerInterval time.Duration
)

func initConfig() {

	// TODO: allow loading config from file, because using env vars is insecure

	viper.BindEnv("log_format_json") //nolint:errcheck
	viper.SetDefault("log_format_json", "true")
	logToJson := viper.GetBool("log_format_json")
	if !logToJson {
		// color console log
		log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stderr})
	}

	viper.BindEnv("log_level") //nolint:errcheck
	viper.SetDefault("log_level", "1")
	logLevel = viper.GetString("log_level")
	level, err := zerolog.ParseLevel(logLevel)
	if err != nil {
		log.Fatal().
			Err(err).
			Str("func", "config"). //TODO: use zerolog stacktrace which also uses func=
			Msg("invalid log_level")
	}
	zerolog.SetGlobalLevel(level)

	viper.BindEnv("debug_decisions_max") //nolint:errcheck
	viper.SetDefault("debug_decisions_max", "-1")
	debugDecisionsMax = viper.GetInt("debug_decisions_max")

	viper.BindEnv("metrics_address") //nolint:errcheck
	viper.SetDefault("metrics_address", ":2112")
	metricsAddr = viper.GetString("metrics_address")

	viper.BindEnv("mikrotik_host") //nolint:errcheck
	mikrotikHost = viper.GetString("mikrotik_host")

	viper.BindEnv("mikrotik_user") //nolint:errcheck
	username = viper.GetString("mikrotik_user")
	if username == "" {
		log.Fatal().
			Str("func", "config").
			Msg("Mikrotik username is not set")
	}

	viper.BindEnv("mikrotik_pass") //nolint:errcheck
	password = viper.GetString("mikrotik_pass")
	if password == "" {
		log.Fatal().
			Str("func", "config").
			Msg("Mikrotik password is not set")
	}

	viper.BindEnv("mikrotik_tls") //nolint:errcheck
	viper.SetDefault("mikrotik_tls", "true")
	useTLS = viper.GetBool("mikrotik_tls")

	viper.BindEnv("mikrotik_ipv4") //nolint:errcheck
	viper.SetDefault("mikrotik_ipv4", "true")
	useIPV4 = viper.GetBool("mikrotik_ipv4")

	viper.BindEnv("mikrotik_ipv6") //nolint:errcheck
	viper.SetDefault("mikrotik_ipv6", "true")
	useIPV6 = viper.GetBool("mikrotik_ipv6")

	viper.BindEnv("mikrotik_firewall_filter_enable") //nolint:errcheck
	viper.SetDefault("mikrotik_firewall_filter_enable", "true")
	enableFirewallFilter = viper.GetBool("mikrotik_firewall_filter_enable")

	viper.BindEnv("mikrotik_firewall_raw_enable") //nolint:errcheck
	viper.SetDefault("mikrotik_firewall_raw_enable", "true")
	enableFirewallRaw = viper.GetBool("mikrotik_firewall_raw_enable")

	viper.BindEnv("mikrotik_address_list") //nolint:errcheck
	viper.SetDefault("mikrotik_address_list", "crowdsec")
	addressList = viper.GetString("mikrotik_address_list")
	if addressList == "" {
		log.Fatal().
			Str("func", "config").
			Msg("mikrotik_address_list cannot be empty")
	}

	if useIPV4 {
		if enableFirewallFilter {
			srcFilterRuleIdsIPv4 = cfgValidateFirewall("ip_firewall_filter_rules_src")
			dstFilterRuleIdsIPv4 = cfgValidateFirewall("ip_firewall_filter_rules_dst")
		}
		if enableFirewallRaw {
			srcRawRuleIdsIPv4 = cfgValidateFirewall("ip_firewall_raw_rules_src")
			dstRawRuleIdsIPv4 = cfgValidateFirewall("ip_firewall_raw_rules_dst")
		}
	}

	if useIPV6 {
		if enableFirewallFilter {
			srcFilterRuleIdsIPv6 = cfgValidateFirewall("ipv6_firewall_filter_rules_src")
			dstFilterRuleIdsIPv6 = cfgValidateFirewall("ipv6_firewall_filter_rules_dst")
		}
		if enableFirewallRaw {
			srcRawRuleIdsIPv6 = cfgValidateFirewall("ipv6_firewall_raw_rules_src")
			dstRawRuleIdsIPv6 = cfgValidateFirewall("ipv6_firewall_raw_rules_dst")
		}
	}

	viper.BindEnv("mikrotik_timeout") //nolint:errcheck
	viper.SetDefault("mikrotik_timeout", "10s")
	timeout = viper.GetDuration("mikrotik_timeout")
	timeoutD, err := time.ParseDuration(timeout.String())
	if err != nil {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("mikrotik_timeout", viper.GetString("mikrotik_timeout")).
			Msg("Failed to parse mikrotik_timeout")
	}

	viper.BindEnv("mikrotik_update_frequency") //nolint:errcheck
	viper.SetDefault("mikrotik_update_frequency", "1h")
	updateFreq = viper.GetDuration("mikrotik_update_frequency")
	updateFreqD, err := time.ParseDuration(updateFreq.String())
	if err != nil {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("mikrotik_update_frequency", viper.GetString("mikrotik_update_frequency")).
			Msg("Failed to parse mikrotik_update_frequency")
	}

	viper.BindEnv("crowdsec_url") //nolint:errcheck
	viper.SetDefault("crowdsec_url", "http://crowdsec:8080/")

	viper.BindEnv("crowdsec_bouncer_api_key") //nolint:errcheck
	crowdsecBouncerAPIKey = viper.GetString("crowdsec_bouncer_api_key")
	if crowdsecBouncerAPIKey == "" {
		log.Fatal().
			Str("func", "config").
			Msg("Crowdsec API key is not set")
	}

	viper.BindEnv("crowdsec_url") //nolint:errcheck
	crowdsecBouncerURL = viper.GetString("crowdsec_url")
	if crowdsecBouncerURL == "" {
		log.Fatal().
			Str("func", "config").
			Msg("Crowdsec LAPI URL is not set")
	}

	viper.BindEnv("crowdsec_origins") //nolint:errcheck
	viper.SetDefault("crowdsec_origins", nil)
	crowdsecOrigins = viper.GetStringSlice("crowdsec_origins")

	viper.BindEnv("default_ttl") //nolint:errcheck
	viper.SetDefault("default_ttl", "3h")
	defaultTTL = viper.GetDuration("default_ttl")
	defaultTTLD, err := time.ParseDuration(defaultTTL.String())
	if err != nil {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("default_ttl", viper.GetString("default_ttl")).
			Msg("Failed to parse default_ttl")
	}

	viper.BindEnv("use_max_ttl") //nolint:errcheck
	viper.SetDefault("use_max_ttl", "false")
	useMaxTTL = viper.GetBool("use_max_ttl")

	viper.BindEnv("default_ttl_max") //nolint:errcheck
	viper.SetDefault("default_ttl_max", "4h")
	maxTTL = viper.GetDuration("default_ttl_max")
	maxTTLD, err := time.ParseDuration(maxTTL.String())
	if err != nil {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("default_ttl_max", viper.GetString("default_ttl_max")).
			Msg("Failed to parse default_ttl_max")
	}
	if maxTTL < updateFreq {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("default_ttl_max", maxTTL.String()).
			Str("mikrotik_update_frequency", viper.GetString("mikrotik_update_frequency")).
			Msg("default_ttl_max can not be shorter than mikrotik_update_frequency")
	}

	viper.BindEnv("trigger_on_update") //nolint:errcheck
	viper.SetDefault("trigger_on_update", "true")
	triggerOnUpdate = viper.GetBool("trigger_on_update")

	viper.BindEnv("ticker_interval") //nolint:errcheck
	viper.SetDefault("ticker_interval", "10s")
	tickerInterval = viper.GetDuration("ticker_interval")
	tickerIntervalD, err := time.ParseDuration(tickerInterval.String())
	if err != nil {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("ticker_interval", viper.GetString("ticker_interval")).
			Msg("ticker_interval value is invalid ")
	}
	if tickerInterval <= 0*time.Second {
		log.Fatal().
			Err(err).
			Str("func", "config").
			Str("ticker_interval", viper.GetString("ticker_interval")).
			Msg("ticker_interval value can not be equal zero or negative")
	}

	all := viper.AllSettings()

	safeConfig := map[string]any{}
	maps.Copy(safeConfig, all)
	safeConfig["mikrotik_pass"] = fmt.Sprintf("%.*s...", 3, password)
	safeConfig["crowdsec_bouncer_api_key"] = fmt.Sprintf("%.*s...", 3, crowdsecBouncerAPIKey)

	for key, val := range safeConfig {
		log.Info().
			Str("func", "config").
			Msgf("Using config: %v=%v", key, val)
	}
	log.Info().
		Str("func", "config").
		Msgf("Setting default TTL to %v", defaultTTLD)
	log.Info().
		Str("func", "config").
		Msgf("Setting max TTL to %v", maxTTLD)
	log.Info().
		Str("func", "config").
		Msgf("Setting mikrotik_timeout to %v", timeoutD)
	log.Info().
		Str("func", "config").
		Msgf("Setting mikrotik_update_requency to %v", updateFreqD)
	log.Info().
		Str("func", "config").
		Msgf("Setting ticker_interval to %v", tickerIntervalD)

}

// cfgValidateFirewall checks if the input string is a valid mikrotik firewall format
// so just numbers and commas
func cfgValidateFirewall(name string) string {

	viper.BindEnv(name) //nolint:errcheck
	value := viper.GetString(name)

	if value == "" {
		log.Fatal().
			Str("func", "config").
			Str(name, value).
			Msgf("%s cannot be empty", name)

	}

	match, _ := regexp.MatchString("^([0-9]+,?)+$", value)
	if !match {
		log.Fatal().
			Str("func", "config").
			Str(name, value).
			Msgf("%s can contain only numbers and commas, aborting", name)
	}

	return value
}
